compiler/qsc_parse/src/prim.rs

```diff
@@ -203,6 +203,17 @@ pub(super) fn recovering<T>(
     }
 }
 
+pub(super) fn recovering_semi(s: &mut Scanner) -> Result<()> {
```

Comment:
nitpick: I wonder if there is a better name for this, because on the one hand it is effectively recovering from a failure to find a token, but on the other hand it does not call `s.recover` so it doesn't actually perform the recovery action. `expect_semi` is not quite right, since that sounds like `expect` in Rust which panics. Maybe something like `check_semi`?

Reply:
There is already a `check_semis` in stmt.rs, and this function is only called from stmt.rs, so maybe it could move there to be next to each other?

Reply:
Yeah naming was tricky. Its semantics are the same as `recovering<T>` and `recovering_token`, in that it attempts to parse an X, and if it can't, it appends an error, and leaves the scanner in a state where we can continue parsing. I agree that it doesn't perform "recovery" per se by eating tokens, but that seems like an implementation detail?

I think `check_semi` sounds too similar to the existing `check_semis` which has totally different behavior. And it does more than checking - it actually parses a semicolon.

I thought about `opt_semi`, but that would evoke the existing `opt()` function which has different semantics as well (doesn't push errors).

I'm not sure, right now my choice is still my favorite but I'm happy to riff on it more.

Reply:
I really think this belongs more next to the other two `recovering*` functions... it's really just a special case of `recovering_token`.